consensus/src/pow/work.rs

```diff
@@ -0,0 +1,163 @@
+#![allow(dead_code)]
+
+use crate::pow::helpers::{
+    calculate_new_target, get_starting_block_time, is_for_retarget, special_rules,
+};
+use crate::pow::temp::BlockIndex;
+use crate::pow::{Error, PoW};
+use common::chain::block::consensus_data::PoWData;
+use common::chain::block::{Block, ConsensusData};
+use common::chain::TxOutput;
+use common::primitives::{Compact, Idable, H256};
+use common::Uint256;
+
+pub fn check_proof_of_work(block_hash: H256, block_bits: Compact) -> Result<bool, Error> {
+    Uint256::try_from(block_bits)
+        .map(|target| {
+            let hash: Uint256 = block_hash.into();
+
+            hash <= target
+        })
+        .map_err(|e| {
+            Error::ConversionError(format!(
+                "conversion of {:?} to Uint256 type: {:?}",
+                block_bits, e
+            ))
+        })
+}
+
+impl PoW {
+    /// The difference (in block time) between the current block and 2016th block before the current one.
+    pub fn actual_timespan(&self, prev_block_blocktime: u32, retarget_blocktime: u32) -> u64 {
+        let actual_timespan = (prev_block_blocktime - retarget_blocktime) as u64;
+
+        num::clamp(
+            actual_timespan,
+            self.min_target_timespan_in_secs(),
+            self.max_target_timespan_in_secs(),
+        )
+    }
+
+    pub fn get_work_required(
+        &self,
+        prev_block_index: &BlockIndex,
+        new_block_time: u32,
+    ) -> Result<Compact, Error> {
+        //TODO: check prev_block_index exists
+        let prev_block_bits = prev_block_index.data.bits();
+
+        if self.no_retargeting() {
+            return Ok(prev_block_bits);
+        }
+
+        let current_height = prev_block_index
+            .height
+            .checked_add(1)
+            .expect("max block height has been reached.");
+
+        let adjustment_interval = self.difficulty_adjustment_interval();
+
+        if is_for_retarget(adjustment_interval, current_height) {
+            let retarget_block_time =
+                get_starting_block_time(adjustment_interval, prev_block_index);
+            self.next_work_required(retarget_block_time, prev_block_index)
+        }
+        // special difficulty rules
+        else if self.allow_min_difficulty_blocks() {
+            Ok(self.next_work_required_for_min_difficulty(new_block_time, prev_block_index))
```

Comment:
Shouldn't the special rule be checked first and if it's false, then adjust the difficulty if needed or am I missing something obvious?

Reply:
I agree with you. In fact, thinking of this, I could argue that this could be a mistake in the implementation. Unless I'm missing something, retargeting has nothing to do with whether min_difficulty is allowed. Testnet also should be allowed to retarget. The way it's done now, it's EITHER testnet has constant difficulty OR a retarget happens, but not both. I'm 90% sure this is not the case in bitcoin. This is why I'd love to have all bitcoin tests here as well to ensure we're doing this right.

Reply:
i have updated the logic, if you guys can check again.

But i cannot do the `CalculationNextRequired` tests from Bitcoin.  
This is equivalent to the `fn next_work_required(...)` but the method is calling another method from a temporary BlockIndex with `todo!()` implementations.

I have added a `TODO` on the mod tests.